src/Microsoft.ML.Core/Environment/HostEnvironmentBase.cs

```diff
@@ -96,12 +96,26 @@ internal interface IMessageSource
     internal abstract class HostEnvironmentBase<TEnv> : ChannelProviderBase, IHostEnvironment, IChannelProvider
         where TEnv : HostEnvironmentBase<TEnv>
     {
+        public bool IsCanceled { get; protected set; }
+
+        private List<IHost> _hosts;
+
+        [BestFriend]
+        internal void CancelExecutionHosts()
+        {
+            foreach (var host in _hosts)
+                if (host is ICancelableHost)
+                    ((ICancelableHost)host).CancelExecution();
+
+            _hosts.Clear();
```

Comment:
>_hosts [](start = 12, length = 6)

Will this class be used from multiple threads? If so then it is not currently threadsafe.

Reply:
Why is it not thread safe? 

---
In reply to: [266555335](https://github.com/dotnet/machinelearning/pull/2797#discussion_r266555335) [](ancestors = 266555335)

Reply:
You can have a race between two instances of CancelExecutionHost - if the _hosts is cleared while the other thread is enumerating you'll get an exception, 
Alternatively there can be a race between CancelExecutionHost() and Register() when item is removed but not cancelled.

---
In reply to: [266558718](https://github.com/dotnet/machinelearning/pull/2797#discussion_r266558718) [](ancestors = 266558718,266555335)